apps/token/src/routes/rewards/epoch-individual-rewards/generate-epoch-individual-rewards-list.ts

```diff
@@ -0,0 +1,88 @@
+import { BigNumber } from '../../../lib/bignumber';
+import { RowAccountTypes } from '../shared-rewards-table-assets/shared-rewards-table-assets';
+import type { RewardFieldsFragment } from '../home/__generated__/Rewards';
+import type { AccountType } from '@vegaprotocol/types';
+
+export interface EpochIndividualReward {
+  epoch: string;
+  rewards: {
+    asset: string;
+    totalAmount: string;
+    rewardTypes: {
+      [key in AccountType]?: {
+        amount: string;
+        percentageOfTotal: string;
+      };
+    };
+  }[];
+}
+
+const emptyRowAccountTypes = Object.keys(RowAccountTypes).map((type) => [
+  type,
+  {
+    amount: '0',
+    percentageOfTotal: '0',
+  },
+]);
+
+export const generateEpochIndividualRewardsList = (
+  rewards: RewardFieldsFragment[]
+) => {
+  // We take the rewards and aggregate them by epoch and asset.
+  const epochsMap = rewards.reduce((map, reward) => {
+    const epochId = reward.epoch.id;
+    const assetName = reward.asset.symbol;
+    const rewardType = reward.rewardType;
+    const amount = reward.amount;
+    const percentageOfTotal = reward.percentageOfTotal;
+
+    // if the rewardType is not of a type we display in the table, we skip it.
+    // @todo check if this is the correct behaviour
+    if (!Object.keys(RowAccountTypes).includes(rewardType)) {
+      return map;
+    }
+
+    if (!map.has(epochId)) {
+      map.set(epochId, { epoch: epochId, rewards: [] });
+    }
+
+    const epoch = map.get(epochId);
+
+    let asset = epoch?.rewards.find((r) => r.asset === assetName);
+
+    if (!asset) {
+      asset = { asset: assetName, totalAmount: '0', rewardTypes: {} };
```

Comment:
question (non-blocking): hmm, wonder if instead of having an empty rewardTypes here, assigning the `rewardTypes: Object.fromEntries(emptyRowAccountTypes)` as it's done below would prevent having to map over the result values again?